lib/kafka/consumer.rb

```diff
@@ -93,16 +100,28 @@ def initialize(cluster:, logger:, instrumenter:, group:, fetcher:, offset_manage
     #   checkpoint.
     # @param max_bytes_per_partition [Integer] the maximum amount of data fetched
     #   from a single partition at a time.
+    # @param refresh_topic_interval [Integer] interval of refreshing the topic list.
+    #   Only work when topic_or_regex is a regex
+    #   If it is 0, the topic list won't be refreshed (default)
+    #   If it is n (n > 0), the topic list will be refreshed every n seconds
     # @return [nil]
-    def subscribe(topic_or_regex, default_offset: nil, start_from_beginning: true, max_bytes_per_partition: 1048576)
+    def subscribe(topic_or_regex, default_offset: nil, start_from_beginning: true, max_bytes_per_partition: 1048576, refresh_topic_interval: 0)
       default_offset ||= start_from_beginning ? :earliest : :latest
 
-      if topic_or_regex.is_a?(Regexp)
-        cluster_topics.select { |topic| topic =~ topic_or_regex }.each do |topic|
-          subscribe_to_topic(topic, default_offset, start_from_beginning, max_bytes_per_partition)
-        end
-      else
+      unless topic_or_regex.is_a?(Regexp)
         subscribe_to_topic(topic_or_regex, default_offset, start_from_beginning, max_bytes_per_partition)
+        return
+      end
+
+      subscribe_to_regex(topic_or_regex, default_offset, start_from_beginning, max_bytes_per_partition)
+      if refresh_topic_interval > 0
+        @thread ||= Thread.new do
+          while true
+            subscribe_to_regex(topic_or_regex, default_offset, start_from_beginning, max_bytes_per_partition) if running?
+            sleep refresh_topic_interval
+          end
+        end
+        @thread.abort_on_exception = true
```

Comment:
This part is problematic. `@thread` is only set on the first invocation of `subscribe`, which can be called multiple times. There should either be a single thread that consults a set of regexes or one thread per regex. Also, the thread or threads need to be killed on consumer shutdown.